lib/oauth2-service.js

```diff
@@ -138,16 +172,55 @@ class OAuth2Service {
     }
 
     const token = this.buildToken(true, xfn, tokenTtl);
-
-    const resp = {
+    const body = {
       access_token: token,
       token_type: 'Bearer',
       expires_in: tokenTtl,
-      scope: req.body.scope,
+      scope,
+    };
+    if (req.body.grant_type !== 'client_credentials') {
+      const credentials = basicAuth(req);
+      const clientId = credentials ? credentials.name : null;
+      body.id_token = this.buildToken(true, (header, payload) => {
+        Object.assign(payload, {
+          sub: 'johndoe',
+          aud: clientId,
+        });
+      }, tokenTtl);
+      body.refresh_token = uuidv4();
+    }
+    const tokenEndpointResponse = {
+      body,
+      statusCode: 200,
     };
+    this.emit('beforeResponse', tokenEndpointResponse);
 
-    return res.json(resp);
+    return res.status(tokenEndpointResponse.statusCode).json(tokenEndpointResponse.body);
   }
-}
 
+  static [authorizeHandler](req, res) {
+    const { scope, state } = req.query;
+    const responseType = req.query.response_type;
+    const redirectUri = req.query.redirect_uri;
+    const code = uuidv4();
+
+    let targetRedirection = `${redirectUri}?code=${code}&scope=${scope}&state=${state}`;
```

Comment:
The arguments in this string needs to be URL-encoded (and everywhere else where you do this.)

Reply:
fixed